ocaml/tests/test_psr.ml

```diff
@@ -0,0 +1,142 @@
+let checkpoint = ref None
+
+let backups = ref None
+
+let default_ps = "default_ps"
+
+let new_ps = "new_ps"
+
+(* we don't make the distinction between a slave & master here.
+   it is intended that whilst the master orchestrates the procedure
+   in the real implementation, the way the master should tell itself
+   to update the PS is no different to the way the master should tell
+   a slave *)
+module Impl = struct
+  type ps = string
+
+  type pss = ps * ps
+
+  type fistpoint = No_fp | Accept_new_ps | Send_new_ps | Finish
+
+  exception Fistpoint of fistpoint
+
+  type host_t = {
+      id: int
+    ; current_ps: string
+    ; staged_ps: string option
+    ; fistpoint: fistpoint
+  }
+
+  type host = host_t ref
+
+  let mk_hosts num =
+    List.init num (fun id ->
+        ref {id; current_ps= default_ps; staged_ps= None; fistpoint= No_fp})
+
+  let save_checkpoint x = checkpoint := Some x
+
+  let retrieve_checkpoint () = !checkpoint
+
+  let backup pss = backups := Some pss
+
+  let retrieve () = Option.get !backups
+
+  let tell_accept_new_ps (_, staged_ps) h =
+    if !h.fistpoint = Accept_new_ps then
```

Comment:
How about
```
match !h.fistpoint with
| Accept_new_ps ->
| .. ->
```

I think this is easier to read and works better with algebraic types.